stats/statsview.cpp

```diff
@@ -80,17 +80,40 @@ void StatsView::mouseReleaseEvent(QMouseEvent *)
 	}
 }
 
+class RootNode : public QSGNode
+{
+public:
+	RootNode();
+	QSGImageNode *imageNode; // imageNode to plot QGRaphicsScene on. Remove in due course.
+	// We entertain one node per Z-level.
+	std::array<QSGNode *, (size_t)ChartZValue::Count> zNodes;
+	std::array<std::vector<ChartItem *>, (size_t)ChartZValue::Count> items;
+};
+
+RootNode::RootNode()
+{
+	for (QSGNode *&zNode: zNodes) {
+		zNode = new QSGNode;
+		appendChildNode(zNode);
+	}
+}
+
 QSGNode *StatsView::updatePaintNode(QSGNode *oldNode, QQuickItem::UpdatePaintNodeData *)
 {
 	// The QtQuick drawing interface is utterly bizzare with a distinct 1980ies-style memory management.
 	// This is just a copy of what is found in Qt's documentation.
-	QSGImageNode *n = static_cast<QSGImageNode *>(oldNode);
-	if (!n)
-		n = rootNode = window()->createImageNode();
+	RootNode *n = static_cast<RootNode *>(oldNode);
```

Comment:
that whole thing is... something. Well designed, Qt people. Well designed.

Reply:
According to the docs, this is designed to be fast, not easy to write for. I fear that the opposite is true: actual applications will be rather inefficient, because it is very hard to get right.

Reply:
I don't doubt that it is __designed__ to be fast. Some of the folks in the Qt team are focused really on the embedded space and there performance on relatively low end silicon is everything.